File: 프로그래머스/Level 2/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

char oper[2] = { '+', '-' };

int numbers_size = 0;
int target_num = 0;
int answer = 0;

void dfs(int idx, int sum, vector<int> numbers) {
    if (idx == numbers_size) {
        if (sum == target_num)
            answer++;
        return ;
    }
    
    for (int i = 0; i < 2; i++) {
        int tmp = sum;
        if (oper[i] == '+')
            tmp += numbers[idx];
        else
            tmp -= numbers[idx];
        
        dfs(idx + 1, tmp, numbers);   
    }
}

int solution(vector<int> numbers, int target) {
    
    numbers_size = numbers.size();
    target_num = target;
    
    dfs(0, 0, numbers);
    
    return answer;
}

File: 프로그래머스/Level 2/타겟 넘버.java
import java.util.*;

class Solution {
    int answer = 0;
    
    public void dfs(int sum, int depth, int target, int[] numbers) {
        if (depth == numbers.length) {
            if (sum == target)
                answer++;
            return ;
        }
        
        dfs(sum + numbers[depth], depth + 1, target, numbers);
        dfs(sum - numbers[depth], depth + 1, target, numbers);
    }
    
    public int solution(int[] numbers, int target) {
      
        dfs(0, 0, target, numbers);
        return answer;
    }
}
